### PR TITLE
Add splash screen

### PR DESCRIPTION
**Pull request by emma-miler**
Code might be a bit shit in places but i hate working with WinAPI.
Automatically disabled on server.
Disabled with `-nosplash`
Automatically hides console on client. Console can be re-enabled by using `-showconsole` or `-nosplash`.

### CODE REVIEW

**Comment by emma-miler** [comment]
Sleep-deprived me PR'ed this while its very much still WIP

**Comment by emma-miler** [comment]
As a quick explanation on why this PR is now `+1800, -1500`, I moved the `include` directory to a shared folder. For more info see: bc48e24dd78cc97b0638460bc982383841eb2f4d

**Comment by emma-miler** [comment]
A quick guide on how to test this PR:
1. Install the build
2. Test by running the game normally. Expected result: there should be a splash screen and no console
3. Test by running the game with `-nosplash`. Expected result: there should be no splash screen and a console, as if this PR was never installed
4. Test by running the game with `-showconsole`. Expected result: there should be both a splash screen and a console
5. Test by running via launchers (VTOL, Viper, FlightCore, etc). Expected result: same as 2.

**Comment by emma-miler** [comment]
For testing custom splash screens, use the example zip provided below. Just add `-altsplash="path_to_folder"`
This can either an absolute path or a path relative to the launcher exe
[customsplash.zip](https://github.com/R2Northstar/NorthstarLauncher/files/10043900/customsplash.zip)

File: NorthstarDLL/hooks.cpp [hunk 1]
```diff
@@ -396,7 +396,15 @@ AUTOHOOK_ABSOLUTEADDR(_LoadLibraryExW, LoadLibraryExW,
...
-	HMODULE moduleAddress = _LoadLibraryExW(lpLibFileName, hFile, dwFlags);
+	HMODULE moduleAddress = NULL;
+	try
+	{
+		moduleAddress = _LoadLibraryExW(lpLibFileName, hFile, dwFlags);
+	}
+	catch (std::exception& e)
+	{
+		spdlog::error(e.what());
+	}
```
Comment (pg9182) [inline NorthstarDLL/hooks.cpp:407]: What's with this?

File: NorthstarDLL/NorthstarDLL.vcxproj [hunk 2]
```diff
@@ -116,13 +116,288 @@
     </PostBuildEvent>
   </ItemDefinitionGroup>
   <ItemGroup>
+    <ClInclude Include="..\include\httplib.h" />
```
Comment (pg9182) [inline NorthstarDLL/NorthstarDLL.vcxproj:119]: Can all these moves be put into a separate PR?

File: NorthstarDLL/NorthstarDLL.vcxproj [hunk 3]
```diff
@@ -102,7 +102,7 @@
       <OptimizeReferences>true</OptimizeReferences>
       <GenerateDebugInformation>true</GenerateDebugInformation>
       <EnableUAC>false</EnableUAC>
-      <AdditionalDependencies>$(ProjectDir)include\MinHook.x64.lib;$(ProjectDir)include\libcurl\lib\libcurl_a.lib;dbghelp.lib;Wldap32.lib;Normaliz.lib;version.lib;%(AdditionalDependencies)</AdditionalDependencies>
+      <AdditionalDependencies>$(SolutionDir)include\MinHook.x64.lib;$(SolutionDir)include\libcurl\lib\libcurl_a.lib;dbghelp.lib;Wldap32.lib;Msimg32.lib;Normaliz.lib;version.lib;%(AdditionalDependencies)</AdditionalDependencies>
```
Comment (pg9182) [inline NorthstarDLL/NorthstarDLL.vcxproj:105]: Why do we need msimg32?

File: NorthstarDLL/logging.cpp [hunk 4]
```diff
@@ -118,16 +120,31 @@ void CustomSink::custom_log(const custom_log_msg& msg)
...
+}
+
 void InitialiseConsole()
 {
-	if (AllocConsole() == FALSE)
-	{
-		std::cout << "[*] Failed to create a console window, maybe a console already exists?" << std::endl;
-	}
-	else
+	bool hasAttached = AttachConsole(-1);
```
Comment (pg9182) [inline NorthstarDLL/logging.cpp:131]: I need to check how this interacts with the docker image before this is merged.

**Review by pg9182 — DISMISSED** [review]

File: NorthstarDLL/NorthstarDLL.vcxproj (hunk 2 above)
Comment (emma-miler) [reply NorthstarDLL/NorthstarDLL.vcxproj:119]: Done. See #337

File: NorthstarDLL/NorthstarDLL.vcxproj (hunk 3 above)
Comment (emma-miler) [reply NorthstarDLL/NorthstarDLL.vcxproj:105]: Needed for `TransparentBlt` function.
https://learn.microsoft.com/en-us/windows/win32/api/wingdi/nf-wingdi-transparentblt

File: NorthstarDLL/hooks.cpp (hunk 1 above)
Comment (emma-miler) [reply NorthstarDLL/hooks.cpp]: Leftover debug statement. Removed now

File: NorthstarDLL/logging.cpp (hunk 4 above)
Comment (emma-miler) [reply NorthstarDLL/logging.cpp:131]: Report back with your findings. More than happy to help if this creates issues.

**Comment by emma-miler** [comment]
Somewhat depends on #337 to be merged first

File: NorthstarDLL/logging.cpp (hunk 4 above)
Comment (emma-miler) [reply NorthstarDLL/logging.cpp:131]: Confirmed working under docker

Comment (pg9182) [reply NorthstarDLL/logging.cpp:131]: Can I see the first 50 or so lines of the log to check a few things?

**Comment by emma-miler** [comment]
I just realised this might not work with `-northstar`. If so, i'll need to make this a separate DLL. Can't test right now, maybe later tonight or tomorrow.

**Comment by emma-miler** [comment]
Soft requirement on https://github.com/R2Northstar/Northstar/pull/395

**Comment by emma-miler** [comment]
Should work with wsock32 launching too now.

**Comment by pg9182** [comment]
Why is the DLL necessary? If resources, why not just embed the splash as-is?

**Comment by emma-miler** [comment]
> Why is the DLL necessary? If resources, why not just embed the splash as-is?

Would rather not duplicate the entire splash screen code into wsock32 as well.

**Comment by emma-miler** [comment]
test

**Review by ASpoonPlaysGames — APPROVED** [review]
Works in testing, using default and a custom splash screen. 
image

**Review by BobTheBob9 — APPROVED** [review]
required changes were discussed previously in vc and made

**Comment by uniboi** [comment]
no way they forgor

File: SplashScreen/dllmain.cpp [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+    case DLL_THREAD_ATTACH:
+    case DLL_THREAD_DETACH:
+    case DLL_PROCESS_DETACH:
+        break;
+    }
+
+	handle = hModule;
+
+    return TRUE;
+}
```
Comment (Jan200101) [inline SplashScreen/dllmain.cpp:24]: indention is a bit screwy

File: SplashScreen/dllmain.cpp [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+                     )
+{
+    switch (ul_reason_for_call)
+    {
+    case DLL_PROCESS_ATTACH:
+    case DLL_THREAD_ATTACH:
+    case DLL_THREAD_DETACH:
+    case DLL_PROCESS_DETACH:
+        break;
+    }
```
Comment (Jan200101) [inline SplashScreen/dllmain.cpp:19]: since we aren't using `ul_reason_for_call`, can this just be removed?

File: NorthstarLauncher/main.cpp [hunk 7]
```diff
@@ -358,9 +367,52 @@ int main(int argc, char* argv[])
...
+				int quote1 = cla.find(quote);
+				int quote2 = (cla.substr(quote1 + 1)).find(quote);
+				altSplash = cla.substr(quote1 + 1, quote2);
+			}
+		}
+		auto splashDLL = LoadLibraryExW(L"splash.dll", 0, LOAD_WITH_ALTERED_SEARCH_PATH);
+		if (!splashDLL)
+		{
+			MessageBoxA(GetForegroundWindow(), "Failed to load splash.dll!", "Northstar Wsock32 Proxy Error", 0);
+			return false;
```
Comment (Jan200101) [inline NorthstarLauncher/main.cpp:398]: Should we really prematurely return just because splashes aren't available?

**Comment by ASpoonPlaysGames** [comment]
@emma-miler is this still going to be worked on?

**Comment by ASpoonPlaysGames** [comment]
Closing due to merge conflicts and probably stale?
